main.cpp

```diff
@@ -272,141 +218,162 @@ struct HeapA
-
-    FloatType& apply( std::function<FloatType&(float&)> func )
+    Numeric& operator+=( Type rhs )
     {
-        if(func)
-            return func(*value);
+        *this->value += rhs;
         return *this;
     }
-    FloatType& apply( void(*funcPtr)(float&) )
+    Numeric& operator-=( Type rhs )
     {
-        if( funcPtr )
-            funcPtr(*value);
+        *this->value -= rhs;
+        return *this;
+    }
+    Numeric& operator*=( Type rhs )
+    {
+        *this->value *= rhs;
         return *this;
     }
 
-private: 
-    float* value = nullptr;
-    FloatType& powInternal(float arg);
-};
-
+    template<class DivType>
+    Numeric& operator/=( const DivType& rhs )
+    {
+        if constexpr ( std::is_same< Type, int >::value )
+        { 
+            if constexpr ( std::is_same< DivType, int >::value )
+            {
+                if ( rhs == 0 )
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+                    return *this;
+                }
+            }
+            else if ( rhs < std::numeric_limits<DivType>::epsilon() )
+            {
+                std::cout << "can't divide integers by zero!" << std::endl;
+                return *this;
+            }
+        }
+        else if ( rhs < std::numeric_limits<Type>::epsilon() )
+        {
+            std::cout << "warning: floating point division by zero!" << std::endl;
+        }
+        *this->value /= rhs;
+        return *this;
+    }
 
-struct DoubleType
-{
-    DoubleType(double nuDouble) : value( new double(nuDouble) ) { }
-    ~DoubleType()
+    Numeric& pow(Type n)
     {
-        delete value;
-        value = nullptr;
+        return powInternal(n);
     }
-   
-   operator double() const {return *value;}
-
-    DoubleType& operator+=( double rhs );
-    DoubleType& operator-=( double rhs );
-    DoubleType& operator*=( double rhs );
-    DoubleType& operator/=( double rhs );
-
-    DoubleType& pow(double d);
-    DoubleType& pow(const IntType& i);
-    DoubleType& pow(const FloatType& f);
-    DoubleType& pow(const DoubleType& d);
-    
-    DoubleType& apply( std::function<DoubleType&(double&)> func )
+
+    Numeric& apply( std::function<Numeric&( std::unique_ptr<Type>& )> func ) //
     {
         if(func)
-            return func(*value);
+            return func(value);
         return *this;
     }
-    DoubleType& apply( void(*funcPtr)(double&) )
+    Numeric& apply( void(*funcPtr)( std::unique_ptr<Type>& ) )
     {
         if( funcPtr )
-            funcPtr(*value);
+            funcPtr(value);
         return *this;
     }
 
-private:
-    double* value = nullptr;
-    DoubleType& powInternal(double arg);
+private: 
+    std::unique_ptr<Type> value;
+    Numeric& powInternal(Type arg)
+    {
+        if(value!= nullptr) 
+            *value = static_cast<Type>(std::pow( *value, arg ));
+        return *this;
```

Comment:
You don't need to check if `value` is a nullptr.  Your constructor prevents that from happening
- [ ] revise